engine/pkg/client/executor_group.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/engine/model"
+	"github.com/pingcap/tiflow/pkg/retry"
+	"github.com/pingcap/tiflow/pkg/security"
+	"go.uber.org/zap"
+)
+
+const (
+	getExecutorBlockTimeout = 10 * time.Second
+)
+
+// ExecutorGroup holds a group of ExecutorClients.
+type ExecutorGroup interface {
+	// GetExecutorClient tries to get the ExecutorClient for the given executor.
+	GetExecutorClient(id model.ExecutorID) (ExecutorClient, bool)
+
+	// GetExecutorClientB tries to get the ExecutorClient for the given executor.
+	// It blocks until either the context has been canceled or the executor ID becomes valid.
+	GetExecutorClientB(ctx context.Context, id model.ExecutorID) (ExecutorClient, error)
+}
+
+// DefaultExecutorGroup is the default implementation for ExecutorGroup.
+type DefaultExecutorGroup struct {
+	mu      sync.RWMutex
+	clients map[model.ExecutorID]ExecutorClient
+
+	logger        *zap.Logger
+	clientFactory executorClientFactory
+}
+
+// NewExecutorGroup creates a new ExecutorGroup.
+func NewExecutorGroup(
+	credentials *security.Credential,
+	logger *zap.Logger,
+) *DefaultExecutorGroup {
+	return newExecutorGroupWithClientFactory(
+		logger,
+		newExecutorClientFactory(credentials, logger))
+}
+
+func newExecutorGroupWithClientFactory(
+	logger *zap.Logger,
+	factory executorClientFactory,
+) *DefaultExecutorGroup {
+	if logger == nil {
+		logger = zap.L()
+	}
+	return &DefaultExecutorGroup{
+		clients:       make(map[model.ExecutorID]ExecutorClient),
+		clientFactory: factory,
+		logger:        logger,
+	}
+}
+
+// GetExecutorClient tries to get the ExecutorClient for the given executor.
+func (g *DefaultExecutorGroup) GetExecutorClient(id model.ExecutorID) (ExecutorClient, bool) {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	if client, ok := g.clients[id]; ok {
```

Comment:
Why not use `return g.clients[id]`?

Reply:
Go does not support this syntax. Refer to https://stackoverflow.com/questions/49921496/passing-on-return-from-a-map-get